### DISCUSSION TITLE
Support for QUALIFY clause?

### DISCUSSION THREAD

**Discussion by r1b** [sqlalchemy/sqlalchemy#12178] [Ideas]
Sample dialect implementations:

- [BigQuery](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#qualify_clause)
- [Snowflake](https://docs.snowflake.com/en/sql-reference/constructs/qualify)
- [Databricks](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-qry-select-qualify.html)
- [DuckDB](https://duckdb.org/docs/sql/query_syntax/qualify.html)
- [Clickhouse](https://clickhouse.com/docs/en/sql-reference/statements/select/qualify)

Seems mostly consistent, except DuckDB which maybe intersects with [named window functions](https://github.com/sqlalchemy/sqlalchemy/issues/8352).

Wondering, should this live in sqlalchemy or is there an extension point for dialects?

**Comment by CaselIT** [comment]
Hi,

While it could be supported by sqlalchemy directly, none of the internal dialects seem to supports it, so it wouldn't be possible to do a round trip test, but compilation would work ok.
Do you know it it's in the roadmap of postgresql or mysql/mariadb?

Regarding the ability for external dialects to add it, I could be possible to have a standalone component, but making it a method of select would probably be a bit of a challenge 

@zzzeek do you have an opinion on this?

**Comment by zzzeek** [comment]
Well I appreciate that this is an entirely novel problem, a SQL keyword that is becoming common but isnt in any of the traditional DBs.

We would need to know more about this SQL keyword.   

1. is it part of the SQL standard? a yes on this would make this easy but I'm not optimistic
2. for all the above DBs, is the syntax of this keyword *exactly the same*?  that is, same word, occurs in the same part of a SELECT statement?   That would make this more easy.  it seems to also be requiring of a window function as the argument

this seems to be a really prominent keyword and it's made it to [Markus Winands blog](https://modern-sql.com/caniuse/qualify) where he states it's not in the SQL standard, but having a compat chart on his site is sort of close to that.

This looked like the kind of thing PostgreSQL would probably add.   But unfortunately not:  https://www.postgresql.org/message-id/1770928.1664862226%40sss.pgh.pa.us   Tom Lane is very opposed to it.     also complicating things is the claim that PG's DISTINCT ON can do the same thing.

One school of thought for me is that I could support going all the way and adding `.qualify()` to `select()`, and just going with it.   ~~I can see PostgreSQL adding support at some point anyway.~~  because it takes in a window function which is an elaborate structure, caching should be done correctly, etc.   But Lane's email is making me hesitate on that.

another school of thought is that maybe we should have a more established path to add new third party clauses to SELECT (and others) with better support for caching and sub-expressions (we are having this problem with MySQL UPDATE/DELETE with LIMIT right now).   QUALIFY here would be very awkward in the current "dialect options" model we have, it would be like ``duckdb_qualify=X, bigquery_qualify=X, snowflake_qualify=X, ...`` which would be ridiculous and repetitive (and will break caching right now anyway).   What would some other plugin system look like?   Well we have `@compiles` also which can do this now, but for what API on `Select` (and how to keep caching working)?      the third party way of doing this would have a lot of "from sqlalchemy_bigquery import select" stuff going on too, also makes it hard to write backend-agnostic code.

this is an odd case will have to think more

**Reply by zzzeek** [reply]
if there was a third party extension on Pypi called sqlalchemy_qualify_select, that is just the `Select` class here, with an appropriate `@compiles` recipe and the caching bits done correctly, that could in fact be cross-platform.

**Reply by zzzeek** [reply]
another idea, make an example of "qualify" and have this be a sqlalchemy.ext with a test suite, and then publish this as "this is how you would make other third party select extensions"  but it's unlikely this would work, people would just ask for everything to be in sqlalchemy.ext

**Reply by CaselIT** [reply]
This may also be tangentially related https://github.com/sqlalchemy/sqlalchemy/issues/8352, not sure if for all engines or only for duckdb

**Reply by zzzeek** [reply]
im thinking of this

```py
from myextension import qualify

stmt = select(A).ext(qualify(partition_by=foobar))

```

on the extension side, mixin:

```py

class StatementExtensions:

    _extensions_traverse_internals = [
        ("_extensions", InternalTraversal.dp_string_clauseelement_dict),
    ]

    @memoized_property
    def _extensions(self):
        return {}

    def ext(self, construct: ExtensionClauseElement):
        raise NotImplementedError()

```

SELECT, UPDATE, DELETE, whatever, includes this:

```py

class Select(...., StatementExtensions, ...):
    _traverse_internals: _TraverseInternalsType = (
        [
              # ...
        ] +  # ...
        + StatementExtensions._extensions_traverse_internals    

     # ...

    def ext(self, construct: ExtensionClauseElement):
        construct.add_to_select(self)

```


Interface:

```py

class ExtensionClauseElement(ClauseElement):
     def add_to_select(self, statement):
          """associate this clauseelement with a SELECT"""
          raise NotImplementedError()

     def add_to_update(self, statement):
          """associate this clauseelement with an UPDATE"""
          raise NotImplementedError()

     def add_to_insert(self, statement):
          """associate this clauseelement with an INSERT"""
          raise NotImplementedError()

     def add_to_delete(self, statement):
          """associate this clauseelement with a DELETE"""
          raise NotImplementedError()

    # ...


```


QUALIFY:

```py

class Qualify(ExtensionClauseElement):
      _traverse_internals = ...  # this has to be a proper traverse internals
     def __init__(self, arg):
           self.arg = coercions ....   # this would be proper coercion, etc.

     def add_to_select(self, select):
         existing = select._extensions.get("org.mydomain.myextension", None)
         if existing is not None:
                self = <somehow combine existing clause with this one, or whatever you want to do>
         select._extensions["org.mydomain.myextension"] = self


@compiles(Qualify)
def _compile_qualify(element, compiler, **kw):
     """generates the string for the QUALIFY clause itself"""
     # ...

# AND THEN, THE TRICKY PART, PLACING IT IN THE SELECT STRING, IM MAKING SOMETHING STUPID UP HERE

@ExtensionClauseElement.places(Select, "duckdb")
@ExtensionClauseElement.places(Select, "snowflakedb")
def _place_qualify(element, element_string, select_string, compiler, **kw):
      return re.replace(r"$|(?=ORDER BY .*)", element_string, select_string)

```

I'll see if I can think more about that last part

**Reply by CaselIT** [reply]
seems like a lot of machinery for something that may not be that common.
unless we have other use cases for this

**Reply by zzzeek** [reply]
we do:

* mysql_update_limit
* mysql_delete_limit
* postgresql ON CONFLICT
* sqlite ON CONFLICT
* mysql WITH DUPLICATE KEY

all of these things would go through this idea (all of which would gain caching support, none of them have caching now)   it would include some standard approach to testing cache keys

**Reply by CaselIT** [reply]
if that's a general approach then great!

**Reply by zzzeek** [reply]
it's one of those big ideas I might decide is too out on a limb by tomorrow

**Comment by mattalbr** [comment]
My google search for this feature led me here, how does this relate to this [documentation](https://docs.sqlalchemy.org/en/21/_modules/examples/syntax_extensions/qualify.html) for qualify as a syntax extension in 2.1? Did this end up getting solved in a generic way as suggested by zzzeek?

**Reply by CaselIT** [reply]
In 2.1 there will be no builtin qualify, but there will be the syntax extension feature, that allows you ho build qualify and similar features.
The idea is that 3rd party dialects will provide such extension to the language, similarly to how distinct_on got re-done using the same syntax extension feature: https://docs.sqlalchemy.org/en/21/dialects/postgresql.html#sqlalchemy.dialects.postgresql.distinct_on

**Comment by r1b** [comment]
Thank you! This is very cool.
